### PR TITLE
Decodable response serialization and major refactor

### PR DESCRIPTION
**Pull request by jshier**
### Issue Link :link:
#2180

### Goals :soccer:
The primary goal of the PR is to add a response serializer that uses `JSONDecoder` to generically decode responses. However, that lead to a major refactor of the response serialization protocols and their implementations.

### Implementation Details :construction:
Functionally, the serializers are the same, they've just been encapsulated into classes, rather than being generated closures. This allows they to be more easily used, and the awkward extensions on `Request` that generated the closures have been removed. Also, the serializations protocols have had their requirements changed from a serialization closure to a throwing function. This allows for far less awkward usage, as well as providing superior semantics with argument labels. 

### Testing Details :mag:
Tests have been added for the decodable serializers and response methods. Additionally, some of the tests have been refactored.

### CODE REVIEW

File: Tests/ResponseTests.swift [hunk 1]
```diff
@@ -323,6 +323,47 @@ class ResponseJSONTestCase: BaseTestCase {
     }
 }
 
+class ResponseJSONDecodableTestCase: BaseTestCase {
```
Comment (jshier) [inline Tests/ResponseTests.swift]: I need more tests here.

File: Tests/ResponseTests.swift [hunk 2]
```diff
@@ -323,6 +323,68 @@ class ResponseJSONTestCase: BaseTestCase {
...
+            expectation.fulfill()
+        }
+
+        waitForExpectations(timeout: timeout, handler: nil)
+
+        // Then
+        XCTAssertNotNil(response?.request)
+        XCTAssertNotNil(response?.response)
+        XCTAssertNotNil(response?.data)
+        XCTAssertNotNil(response?.data)
```
Comment (mRs-) [inline Tests/ResponseTests.swift]: you are testing two times the data, is this intended or a typo?

Comment (jshier) [reply Tests/ResponseTests.swift]: No, I must have copy pasted something.

Comment (jshier) [reply Tests/ResponseTests.swift]: Removed in [9e374bf](https://github.com/Alamofire/Alamofire/pull/2265/commits/9e374bfdb2cd81577d89338025ced09efc4608c5).

**Pull request merged**